src/software/backend/simulation/physics_ball.cpp

```diff
@@ -0,0 +1,51 @@
+#include "software/backend/simulation/physics_ball.h"
+
+#include "shared/constants.h"
+#include "software/backend/simulation/box2d_util.h"
+
+PhysicsBall::PhysicsBall(std::shared_ptr<b2World> world, const Ball &ball)
+{
+    // All the BodyDef must be defined before the body is created.
+    // Changes made after aren't reflected
+    ball_body_def.type = b2_dynamicBody;
+    ball_body_def.position.Set(ball.position().x(), ball.position().y());
+    //    ball_body_def.linearVelocity.Set(ball.velocity().x(), ball.velocity().y());
+    // The ball can potentially move relatively quickly, so treating it as a "bullet"
+    // helps prevent tunneling and other collision problems
+    // See the "Breakdown of a collision" section of:
+    // https://www.iforce2d.net/b2dtut/collision-anatomy
+    //    ball_body_def.bullet = true;
+    ball_body = world->CreateBody(&ball_body_def);
+    b2Vec2 velocity_impulse_direction(ball.velocity().x(), ball.velocity().y());
+    ball_body->ApplyLinearImpulse(velocity_impulse_direction, ball_body->GetWorldCenter(),
+                                  true);
+
+    ball_shape.m_radius = BALL_MAX_RADIUS_METERS;
+
+    ball_fixture_def.shape = &ball_shape;
+    // These values do not reflect the ball in real life and may still need some tuning.
+    // For now, they are "ideal" values that give the ball perfectly elastic collision
+    // and no friction
+    // TODO: tune values if necessary (#768)
+    ball_fixture_def.density     = 1.0;
+    ball_fixture_def.restitution = 1.0;
+    ball_fixture_def.friction    = 0.0;
+
+    ball_body->CreateFixture(&ball_fixture_def);
+}
+
+Ball PhysicsBall::getBallWithTimestamp(const Timestamp &timestamp) const
+{
+    auto position = Point(ball_body->GetPosition().x, ball_body->GetPosition().y);
+    auto velocity =
+        Vector(ball_body->GetLinearVelocity().x, ball_body->GetLinearVelocity().y);
+    return Ball(position, velocity, timestamp);
+}
+
+void PhysicsBall::removeFromWorld(std::shared_ptr<b2World> world)
+{
+    if (bodyExistsInWorld(ball_body, world))
+    {
+        world->DestroyBody(ball_body);
+    }
+}
```

Comment:
This might cause problems if we try to destroy a body mid simulation. I think Box2D locks the world during its timesteps so we can't edit it.

Reply:
wow, nice catch.

Reply:
@jared-paul locks the works as-in with a mutex? Or just completely prevents changes from happening? I'd be surprised that it exposes this as an API if it wouldn't work. Can you explain more?

Reply:
I don't know how it locks it, I just know it does during timestep executions. For example, if you were to destroy a body in the middle of a collision that would be bad, so Box2D will throw a "world IsLocked" exception. This might not be the behaviour now, I just know that it used to be like that, they might've updated it since then.

https://www.iforce2d.net/b2dtut/removing-bodies here's a good source

Reply:
Ah ok I understand, good catch. For now this removal won't happen during a collision callback or during a physics step (anything that might need callbacks is TBD and needs to be set up). This should be safe for now but I'll definitely keep this in mind going forwards.

Sound good?

Reply:
Yep!

Quick thing: should we make a "PhysicsWorld" abstraction like we did with the ball? so instead of having "removeFromWorld" in "PhysicsBall" we could have "removeBall()" in "PhysicsWorld"

Reply:
I'll get there eventually, trying to build it up one class at a time so PRs aren't huge. Will definitely be revising stuff as I go.

Reply:
@jared-paul just need your approval then